File: test/bench/test_bv.cpp
/*
 * Copyright (c) 2023 - 2026 Chair for Design Automation, TUM
 * Copyright (c) 2025 - 2026 Munich Quantum Software Company GmbH
 * All rights reserved.
 *
 * SPDX-License-Identifier: MIT
 *
 * Licensed under the MIT License
 */

#include "bench/BV.hpp"
#include "bench/Evaluation.hpp"

#include "support/Diagnostics.hpp"
#include "support/TestSupport.hpp"

#include "gtest/gtest.h"

#include <string>

namespace {

using mqt::bench::BV;
using mqt::bench::BVMethod;
using mqt::bench::BVOptions;
using mqt::bench::Output;

TEST(BV, UsesTheStaticMethodByDefault) {
  const auto benchmark =
      ::mqt::test::value(BV::create({.hiddenBitstring = "101"}));
  EXPECT_EQ(benchmark.options().method, BVMethod::Static);
  EXPECT_EQ(benchmark.output(), (Output{"result", 3}));
}

TEST(BV, ValidatesTheConfiguredInstance) {
  // NOLINTNEXTLINE(clang-analyzer-optin.core.EnumCastOutOfRange)
  constexpr auto invalidMethod = static_cast<BVMethod>(2);
  EXPECT_EQ(::mqt::test::errorKind(
                [&] { return BV::create({.hiddenBitstring = ""}); }),
            ::mqt::ErrorCategory::InvalidArgument);
  EXPECT_EQ(::mqt::test::errorKind(
                [&] { return BV::create({.hiddenBitstring = "10x"}); }),
            ::mqt::ErrorCategory::InvalidArgument);
  EXPECT_EQ(::mqt::test::errorKind([&] {
              return BV::create({.hiddenBitstring = std::string(
                                     BVOptions::MAX_BITS + 1, '0')});
            }),
            ::mqt::ErrorCategory::InvalidArgument);
  EXPECT_EQ(::mqt::test::errorKind([&] {
              return BV::create(
                  {.hiddenBitstring = "1", .method = invalidMethod});
            }),
            ::mqt::ErrorCategory::InvalidArgument);
}

TEST(BV, GivesTheHiddenBitstringAsASelectedOutcome) {
  for (const auto method : {BVMethod::Static, BVMethod::Dynamic}) {
    const auto benchmark = ::mqt::test::value(
        BV::create({.hiddenBitstring = "101", .method = method}));
    EXPECT_DOUBLE_EQ(::mqt::test::value(benchmark.probability("101")), 1.);
    EXPECT_DOUBLE_EQ(::mqt::test::value(benchmark.probability("011")), 0.);

    const auto evaluation =
        ::mqt::test::value(benchmark.evaluate({{"101", 80}, {"011", 20}}));
    EXPECT_DOUBLE_EQ(evaluation.totalVariationDistance, 0.2);
    EXPECT_DOUBLE_EQ(evaluation.squaredHellingerFidelity, 0.8);
    ASSERT_TRUE(evaluation.successProbability);
    EXPECT_DOUBLE_EQ(*evaluation.successProbability, 0.8);
  }
}

} // namespace

File: test/python/bench/test_bv.py
# Copyright (c) 2023 - 2026 Chair for Design Automation, TUM
# Copyright (c) 2025 - 2026 Munich Quantum Software Company GmbH
# All rights reserved.
#
# SPDX-License-Identifier: MIT
#
# Licensed under the MIT License

"""Python bindings for the bv benchmark."""

from __future__ import annotations

import pytest

from mqt.core.bench import bv

from .utils import assert_generates

METHODS = (bv.Method.STATIC, bv.Method.DYNAMIC)


def _make_benchmark(method: bv.Method) -> bv.BV:
    return bv.BV(bv.Options(hidden_bitstring="101", method=method))


@pytest.mark.parametrize("method", METHODS)
def test_bv_reference(method: bv.Method) -> None:
    """Use the hidden bitstring as the deterministic reference result."""
    assert _make_benchmark(method).probability("101") == 1


@pytest.mark.parametrize("method", METHODS)
def test_bv_evaluation(method: bv.Method) -> None:
    """Report successful samples for both methods."""
    assert _make_benchmark(method).evaluate({"101": 10}).success_probability == 1


@pytest.mark.parametrize("method", METHODS)
def test_bv_manifest_roundtrip(method: bv.Method) -> None:
    """Preserve the benchmark identity through its manifest."""
    benchmark = _make_benchmark(method)
    assert bv.BV.from_manifest_json(benchmark.manifest_json).case_id == benchmark.case_id


@pytest.mark.parametrize("method", METHODS)
def test_bv_generation(method: bv.Method) -> None:
    """Generate both Bernstein--Vazirani methods."""
    assert_generates(_make_benchmark(method).generate())
